fix: (Core) Overflowing, stacked and collapsible tabs

salarenko opened the pull request:
#### Please provide a link to the associated issue.
Closes: #1410 

#### Please provide a brief summary of this pull request.
This PR:
- Adds collapsible tabs functionality
- Adds collapsible tabs overflow functionality
- Adds maximum number of visible tabs functionality
- Adds adds stacked tabs content functionality
- Adds `getElementCapacity` and `getElementWidth` utils for counting elements capacity and width (with margins)
- Adds a `[scrollSpyDisabled]` attribute to disable `ScrollSpy`
- Abandons tab index based API

#### BREAKING CHANGES:
- `selectedIndex ` input has been removed, use `TabPanelComponent.open()`
- `selectedIndexChange` output has been removed, use `selectedTabChange`
- `TabListComponent.selectTab` has been removed, use `TabPanelComponent.open()`
- `TabPanelComponent.triggerExpandedPanel()` has been removed, use `TabPanelComponent.open()`

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

InnaAtanasova commented on libs/core/src/lib/tabs/tab-item/tab-item.directive.ts:
```diff
@@ -18,12 +18,9 @@ export type TabItemState = 'success' | 'error' | 'warning' | 'information' | 'ne
 })
 export class TabItemDirective implements CssClassBuilder, OnChanges, OnInit {
-    /** Apply user custom styles */
+    /** @hidden Apply user custom styles */
```
Why do you hide it? If it's hidden it should be prefixed with `_`

salarenko replied:
Change reverted

InnaAtanasova commented on libs/core/src/lib/tabs/tab-utils/tab-info.class.ts:
```diff
@@ -0,0 +1,63 @@
+
+    get id(): string {
+        return this.panel.id;
+    }
+
+    get ariaLabel(): string {
+        return this.panel.ariaLabel;
+    }
+
+    get ariaLabelledBy(): string {
+        return this.panel.ariaLabelledBy;
+    }
+
+    get disabled(): boolean {
+        return this.panel.disabled;
+    }
+
+    get isHeader(): boolean {
+        return this.panel.header;
+    }
+
+    get tabState(): TabItemState {
+        return this.panel.tabState;
+    }
+
+    get titleTemplate(): TemplateRef<any> {
+        return this.panel.titleTemplate;
+    }
+
+    get count(): string {
+        return this.panel.count;
+    }
+
+    get title(): string {
+        return this.panel.title;
+    }
+
+    get glyph(): string {
+        return this.panel.glyph;
+    }
```
Missing the typedocs

JKMarkowski replied:
Should it even be included into typedocs?

salarenko replied:
Added documentation annotations

salarenko replied:
Typedocs added

InnaAtanasova commented:
<img width="674" alt="Screen Shot 2020-12-16 at 5 20 04 PM" src="https://user-images.githubusercontent.com/39598672/102414194-11d14780-3fc4-11eb-8eb1-a288ae71923b.png">

On resizing I noticed that the `More` button "jumps" and goes on the second line.

InnaAtanasova commented:
I wrote in the previous PR that the Overflow button ("More") is not customizable. It could be anything or just an arrow and needs to handle the i18n.

JKMarkowski commented on libs/core/src/lib/tabs/tab-list.component.html:
```diff
@@ -54,11 +57,41 @@
     </ng-container>
+
+    <li class="fd-tabs__item fd-tabs__item--overflow" #overflowTrigger *ngIf="_isCollapsed">
+        <a class="fd-tabs__link" [attr.tabindex]="0" [fdMenuTrigger]="menu">
```
There is no way to trigger menu, by keyboard 
![image](https://user-images.githubusercontent.com/26483208/102492883-9b3f5300-4072-11eb-9e30-95bdcbcdc192.png)

JKMarkowski commented on libs/core/src/lib/tabs/tab-list.component.ts:
```diff
@@ -90,106 +152,299 @@ export class TabListComponent implements AfterViewInit, OnChanges, OnDestroy {
+    }
+
+    /** @hidden */
+    private _listenOnTabPanelsAndInitiallyExpandTabPanel(): void {
+        this._tabPanelsChange$.pipe(
+            map(_ => this._tabArray.find(tab => tab.active && !tab.disabled)),
+            filter(tab => !tab),
+            map(_ => this._tabArray.find(tab => !tab.disabled)),
+            filter(tab => !!tab),
+            delay(0)
+        ).subscribe(tab => this._expandTab(tab.panel, true));
     }
```
How is it going to work? There are 2 filters, that are excluding eachother

salarenko replied:
This is working well:
- first filter check whether there is a active tab
- second filter check whether there is a tab that might be activated

Actually this can be simplified to:
```
        this._tabPanelsChange$.pipe(
            filter(_ => !this._tabArray.some(tab => tab.active)),
            map(_ => this._tabArray.find(tab => !tab.disabled)),
            filter(Boolean),
            delay(0)
        ).subscribe(tab => this._expandTab(tab.panel, true));
```

JKMarkowski reviewed (changes requested):
Great Job! I left some comments regarding code / keyboard support,

salarenko commented:
@InnaAtanasova Thanks for the review.
- "More v" tab not fitting into the tab panel has been fixed
- Expand tab text can be customized, but info in the docs was missing. I've improved the docs.
- Added required doc annotation changes

Except that I've also moved expand trigger to separate component.

InnaAtanasova commented:
Looks good to me. Thanks for addressing the comments. Awesome job!

InnaAtanasova commented:
P.S. I can still reproduce the issue with the More button going on second row when resizing the screen.

salarenko commented:
@InnaAtanasova @JKMarkowski 

[This commit](https://github.com/SAP/fundamental-ngx/pull/4120/commits/b9569ea7bd9180f4433b6bcca5112bcfb211ec81) has introduced additional wrapper on tabs. Now when tabs cannot fit when resizing they will overflow without moving the "More v" tab to second row.

This diverge from Fundamental Styles structure.

mikerodonnell89 commented:
Nicely done! Couple very minor issues.  I think enter key/space bar when "more" is focused should open the overflow popover.  Also the icon/text needs some spacing on the more button in RTL mode

![Screen Shot 2020-12-18 at 2 13 02 PM](https://user-images.githubusercontent.com/2471874/102662024-369bfb00-413b-11eb-9d8b-9ecfe6a9931f.png)

When first scrolling the stacked example, the selected tab does not update.  But it does on subsequent scrolls. (wondering if this issue would also be present with stacking wizard if all steps are rendered stacked on first load)

![tabstackscroll](https://user-images.githubusercontent.com/2471874/102662690-8d560480-413c-11eb-9329-d9d961e92d07.gif)

salarenko commented:
@mikerodonnell89 Thanks for the review.
- I've improved RTL
- Added keyboard support for expand trigger

JKMarkowski reviewed (approved):
LGTM

stefanoScalzo commented:
Could u add to the programmatic selection example?

mikerodonnell89 commented:
Pushed a fix for the bug I previously mentioned with proper tab not being highlighted when first scrolling stacked content
